[Malcolm Low] iP

mlzt2000 opened the pull request:
### Duke

> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes))

Duke is a chatbot that helps you manage your day to day tasks.

Have you ever encountered these problems in your life:
- [ ] Too many things to do
- [ ] Not sure when you need to do them
- [ ] ~~Too lazy to do them~~ (Sorry Duke can't solve that for you! 😝)

**Then you need to try Duke!**

It can:
1. Add 3 types of tasks: Todos, Deadlines, and Events
2. List all your tasks
3. Mark tasks as completed or incomplete
4. Delete tasks

Best of all, it is text-based and _super easy to learn_!

Here's how you can get it
- Visit [this link](https://www.example.org/)
- Download Duke.jar
- Double click it to run.
- Start using it to manage your tasks! 😄 

Here's a small snippet of the `code` that runs Duke:
```java
public static void main(String[] args) {
    new Duke("Duke.txt").run();
}
```

Commit:
Add Gradle support

mlzt2000 committed:
Update Duke.java

Level-1

mlzt2000 committed:
Level 2

mlzt2000 committed:
level 3

mlzt2000 committed:
level 4

mlzt2000 committed:
A-TextUITesting

mlzt2000 committed:
level-5

mlzt2000 committed:
level-6

mlzt2000 committed:
A-Enums

mlzt2000 committed:
Level-7

deeyonn commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,214 @@
+    private static final String TAB = "    ";
+    private static final String LINE = String
+            .format("%s%s", TAB, "____________________________________________________________");
+    private static final String WELCOMEMESSAGE = String
+            .format("Hello! I'm Duke\n%s What can I do for you?", TAB);
+    private static final String ENDMESSAGE = "Bye. Hope to see you again soon!";
+    private static final String CREATEMESSAGE = "Got it. I've added this task:";
+    private static final String MARKMESSAGE = "Nice! I've marked this task as done:";
+    private static final String UNMARKMESSAGE = "OK, I've marked this task as not done yet:";
+    private static final String LISTMESSAGE = "Here are the tasks in your list:";
+    private static final String DELETEMESSAGE = "Noted. I've removed this task:";
+
+    // Commands
+    private static final String ENDCOMMAND = "bye";
+    private static final String PRINTCOMMAND = "list";
+    private static final String MARKCOMMAND = "mark";
+    private static final String UNMARKCOMMAND = "unmark";
+    private static final String TODOCOMMAND = "todo";
+    private static final String DEADLINECOMMAND = "deadline";
+    private static final String EVENTCOMMAND = "event";
+    private static final String DELETECOMMAND = "delete";
```
I noticed that you have a lot of contant variables, consider the following requirements below:
`Constant names must be all uppercase using underscore to separate words.`
`Associated constants should have a common prefix.`

deeyonn commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,214 @@
+    private static final String DELETECOMMAND = "delete";
+
+    // Data structures
+    private static final List<Task> list = new ArrayList<>();
```
Since `list` is used to store a few instances of `Task`, consider using plural forms to describe what is it you're storing in the list:
`Plural form should be used on names representing a collection of objects.`

deeyonn commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,214 @@
+     */
+    private static void markTask(String index) {
+        index = index.strip();
+        int i = Integer.parseInt(index);
```
I'm not sure if a change here is required, but consider the following requirements:
`Variables named j, k etc. should be used for nested loops only.`

deeyonn commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,214 @@
+            try {
+                String inputCmd = sc.next();
+                String inputRem = sc.nextLine();
+                switch (inputCmd) {
+                    case (ENDCOMMAND):
+                        canNext = false;
+                        break;
+                    case (PRINTCOMMAND):
+                        printAll();
+                        break;
+                    case (MARKCOMMAND):
+                        markTask(inputRem);
+                        break;
+                    case (UNMARKCOMMAND):
+                        unmarkTask(inputRem);
+                        break;
+                    case (TODOCOMMAND):
+                        createTask(inputRem, TaskType.TODO);
+                        break;
+                    case (DEADLINECOMMAND):
+                        createTask(inputRem, TaskType.DEADLINE);
+                        break;
+                    case (EVENTCOMMAND):
+                        createTask(inputRem, TaskType.EVENT);
+                        break;
+                    case (DELETECOMMAND):
+                        deleteTask(inputRem);
+                        break;
+                    default: throw new DukeException("☹ OOPS!!! I'm sorry, but I don't know what that means :-(");
```
I see that you're making good use of switch statements! Consider the following requirements:
`The switch statement should have the following form: Note there is no indentation for case clauses.`

deeyonn reviewed (commented):
Great job so far! 😄 Some points to note:
- Files can be packaged to help with keeping things tidy.
- Consider referencing the coding standard [requirements](https://se-education.org/guides/conventions/java/intermediate.html) for CS2103

mlzt2000 committed:
Level-7

mlzt2000 committed:
Level-8

mlzt2000 committed:
Level-8

mlzt2000 committed:
Level-8

mlzt2000 committed:
Merge branch 'branch-Level-8'

mlzt2000 committed:
Merged Level-7 and Level-8 into master

ish1506 commented on src/main/java/ip/Deadline.java:
```diff
@@ -0,0 +1,28 @@
+import java.time.LocalDateTime;
+import java.time.format.DateTimeFormatter;
+
+class Deadline extends Task {
```
Remember to add header comments to all public classes and methods! I noticed this issue in other places too.

ish1506 commented on src/main/java/ip/Duke.java:
```diff
@@ -0,0 +1,287 @@
+import java.util.List;
+import java.util.Scanner;
+
+public class Duke {
```
Perhaps you could extract out `Ui`, `Storage` and `Parser` classes to handle the different functionalities?

ish1506 commented on src/main/java/ip/Duke.java:
```diff
@@ -0,0 +1,287 @@
+
+    private static void loadFile() {
+        try {
+            // Create Duke.txt if it doesn't exist.
```
I like the comments which explain the code snippets.

ish1506 commented on src/main/java/ip/Duke.java:
```diff
@@ -0,0 +1,287 @@
+            String description = "";
+            String extraInfo = "";
+            switch(type) {
+                case TODO:
```
Remember to follow the module coding standards! Easily fixed with the IDE configurations.
```suggestion
            switch (type) {
            case TODO:
```

jolynloh commented on src/main/java/ip/Deadline.java:
```diff
@@ -0,0 +1,28 @@
+
+class Deadline extends Task {
+
+    protected LocalDateTime by;
```
Perhaps using `deadline` instead of `by` would be more intuitive as a standalone name? 
Correspondingly, the getter method could be phrased as `getDeadline()`

According to the code quality guidelines, 
> "Use Nouns for Things and Verbs for Actions"

ish1506 commented on src/main/java/ip/Task.java:
```diff
@@ -0,0 +1,40 @@
+    protected Task(String description, boolean isDone) {
+        this.description = description;
+        this.isDone = isDone;
+    }
```
Perhaps you can simplify your constructors by marking `isDone` to be `false` by default.
```suggestion
    protected boolean isDone = false;

    Task(String description) {
        this.description = description;
    }
    
    protected Task(String description, boolean isDone) {
        this(description);
        this.isDone = isDone;
    }
```

jolynloh commented on src/main/java/ip/Duke.java:
```diff
@@ -0,0 +1,287 @@
+            // Create Duke.txt if it doesn't exist.
+            File f = new File(STORAGE_FILE_PATH);
+            if (!f.exists()) {
+                boolean fileCreated = f.createNewFile();
```
Consider using a name that sounds like a boolean, such as `isFileCreated`

jolynloh commented on src/main/java/ip/Duke.java:
```diff
@@ -0,0 +1,287 @@
+            boolean isDone = isDoneStr.equals("X");
+
+            Task newTask;
+            switch (taskTypeStr) {
+                case "T":
+                    newTask = new ToDo(description, isDone);
+                    break;
+                case "D":
+                    String by = taskArray[3];
+                    newTask = new Deadline(description, by, isDone);
+                    break;
+                case "E":
+                    String at = taskArray[3];
+                    newTask = new Event(description, at, isDone);
+                    break;
+                default:
+                    throw new DukeException(String.format("Could not load task: %s", description));
+            }
```
Good use of switch case here! Do consider the following requirements:
> "The switch statement should have the following form: Note there is no indentation for case clauses."

jolynloh commented on src/main/java/ip/Duke.java:
```diff
@@ -0,0 +1,287 @@
+                    throw new DukeException("Type of task is not recognised!");
+            }
+            taskList.add(newTask);
+            String singulStr = "Now you have 1 task in the list";
```
Perhaps using a standard word for the String variable name would make it easier to interpret its value and adhere more closely to the following requirement? 
> "Use correct spelling in names. Avoid 'texting-style' spelling."

I noticed the same issue in one or two other places too

jolynloh commented on src/main/java/ip/Event.java:
```diff
@@ -0,0 +1,30 @@
+
+class Event extends Task {
+
+    protected LocalDateTime at;
```
Perhaps using a noun such as `time` would be more intuitive than a preposition?

jolynloh commented on src/main/java/ip/Duke.java:
```diff
@@ -0,0 +1,287 @@
+    }
+
+    private static void prettyPrint(String printable) {
+        String out = String.format("%s\n%s %s\n%s", LINE, TAB, printable, LINE);
```
Perhaps using a noun for variable names would be better able to capture its value/meaning? I noticed the same issue in several other places too

jolynloh reviewed (commented):
Looks good, overall pretty close adherence to coding quality guidelines except at a few places! 👍

ish1506 commented on src/main/java/ip/Duke.java:
```diff
@@ -0,0 +1,287 @@
+        String inputRem = sc.nextLine();
+        String storageStr = String.format("%s %s", inputCmd, inputRem);
+        switch (inputCmd) {
+        case (END_COMMAND):
```
I think brackets aren't needed here.

```suggestion
        case END_COMMAND:
```

ish1506 commented on src/main/java/ip/Duke.java:
```diff
@@ -0,0 +1,287 @@
+        int i = Integer.parseInt(index);
+        i--;
+        Task task = taskList.remove(i);
+        String singulStr = "Now you have 1 task in the list";
```
I think the variable name could be clearer here.

ish1506 reviewed (commented):
Code quality guidelines were largely adhered to, other than some nits that affected readability.

mlzt2000 committed:
A-MoreOOP

mlzt2000 committed:
A-Packages

mlzt2000 committed:
A-Gradle

mlzt2000 committed:
Merge branch 'add-gradle-support'

mlzt2000 committed:
A-JUnit

mlzt2000 committed:
A-Jar

mlzt2000 committed:
Modify erroneous packaging in Duke.java

mlzt2000 committed:
Add ability to find keywords

Find command: add find command
and supporting methods.

Find command did not exist.

Finding specific tasks can be useful to
the user to identify things they need to
do if their list of tasks is long.

Let's:
- Allow the Parser to identify the find
command.
- Allow the TaskList to search tasks for keywords.
- Allow the Task to support finding keywords in its description.
- Allow UI to display the appropriate message.

mlzt2000 committed:
Updated javadoc for classes

Classes and methods did not have documentation

Documented classes and methods make it easier for others to
understand the code.

Let's add documentation that describes what each class and method is
responsible for doing.

mlzt2000 committed:
Modified handleNext() in Parser.java

handleNext() currently separates the command from the rest of the user
input.

This violates the single layer of abstraction principle.

Let's abstract this process by creating helper functions to perform this.

mlzt2000 committed:
Merge branch 'branch-A-JavaDoc'

mlzt2000 committed:
Merge branch 'branch-Level-9'

mlzt2000 committed:
Removed Duke Package

mlzt2000 committed:
Added ability to run checkstyles

Currently, checking of coding conventions is done manually.

This task can be made more efficient by automating it.

Let's add the ability for Gradle to run checkstyle automatically, and
include the necessary dependencies.

mlzt2000 committed:
Added support for JavaFX

mlzt2000 committed:
Added support for JavaFX and GUI

mlzt2000 committed:
Added assertions

Code previously did not utilise assertions.

Using assertions can ensure that the data that the code works with is
consistent with the assumptions made about the data.

Let's add assertions to methods to ensure that input parameters used are
not null.

mlzt2000 committed:
Removed Checkstyle resources

mlzt2000 committed:
Added support for Checkstyles

mlzt2000 committed:
Refactored Storage.java

mlzt2000 committed:
Modified TaskList and UI to use Streams

Current implementation of TaskList does not utilise Streams to manage
collections.

Using Streams makes it easier to understand what the purpose of the
code is because the code is cleaner.

Let's utilise Streams to support finding and printing tasks.

mlzt2000 committed:
Merge pull request #2 from mlzt2000/branch-A-Streams

Modified TaskList and UI to use Streams

mlzt2000 committed:
Merge branch 'master' into branch-A-Assertions

mlzt2000 committed:
Merge branch 'master' into branch-A-CodeQuality

mlzt2000 committed:
Merge pull request #3 from mlzt2000/branch-A-Assertions

Branch a assertions

mlzt2000 committed:
Merge branch 'master' into branch-A-CodeQuality

mlzt2000 committed:
Merge pull request #4 from mlzt2000/branch-A-CodeQuality

Branch a code quality

mlzt2000 committed:
Added ability to tag Tasks and find tags.

Tasks currently cannot be tagged by the user.

Tagging tasks can help the user to classify Tasks more easily, and filter
Tasks based on their tags.

Let's:
- allow Tasks to store tags
- allow Tasks to add tags
- allow Duke to filter Tasks based on the tags

mlzt2000 committed:
Updated user guide

mlzt2000 committed:
Modified application path in build.gradle

mlzt2000 committed:
Bug fixes

mlzt2000 committed:
Bug fixes

mlzt2000 committed:
Bug fixes
